### PR TITLE
locations api exposed

### PR DESCRIPTION
**Pull request by kayumuzzaman**

### CODE REVIEW

File: packages/gateway/src/features/restLocation/locationHandler.ts [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+  const partOfLocation = payload.partOf.split('/')[1]
+
+  const locations = await getLocationsByIdentifier(
+    `ADMIN_STRUCTURE_${String(payload.statisticalID)}`
+  ).catch((err) => {
+    throw err
+  })
+
+  if (locations.length !== 0) {
+    throw conflict(`statisticalID ${payload.statisticalID} is already exist`)
```
Comment (Zangetsu101) [inline packages/gateway/src/features/restLocation/locationHandler.ts:167]:
```suggestion
    throw conflict(`statisticalID ${payload.statisticalID} already exists`)
```

File: packages/gateway/src/features/restLocation/locationHandler.ts [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  }
+
+  if (
+    partOfLocation !== '0' &&
+    (partOfLocation !== null || partOfLocation !== '')
```
Comment (Zangetsu101) [inline packages/gateway/src/features/restLocation/locationHandler.ts:172]:
```suggestion
    Boolean(partOfLocation)
```

File: packages/gateway/src/features/restLocation/locationHandler.ts [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  district: string
+  state: string
+}
+
+type UpdateLocation = {
+  name: string
+  alias: string
+  status: string
+  statistics: LocationStatistic
+}
```
Comment (Zangetsu101) [inline packages/gateway/src/features/restLocation/locationHandler.ts:86]: these should be optional right?

File: packages/gateway/src/features/restLocation/locationHandler.ts [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+const locationStatisticSchema = Joi.object({
+  year: Joi.number(),
+  male_population: Joi.number(),
+  female_population: Joi.number(),
+  population: Joi.number(),
+  crude_birth_rate: Joi.number()
+})
```
Comment (Zangetsu101) [inline packages/gateway/src/features/restLocation/locationHandler.ts:94]: We should make these required as well.

**Review by Zangetsu101 — CHANGES_REQUESTED** [review]

File: packages/gateway/src/constants.ts [hunk 5]
```diff
@@ -14,6 +14,8 @@ export const HOST = process.env.HOST || '0.0.0.0'
 export const PORT = process.env.PORT || 7070
 export const HOSTNAME = process.env.DOMAIN || '*'
 export const FHIR_URL = process.env.FHIR_URL || 'http://localhost:5001/fhir'
+export const HEARTH_URL = process.env.HEARTH_URL || 'http://localhost:3447/fhir'
+export const ORG_URL = 'http://opencrvs.org'
```
[the remark this answers is not in the document]
Comment (kayumuzzaman) [reply packages/gateway/src/constants.ts]: added to docker compose

File: packages/gateway/src/features/fhir/utils.ts [hunk 6]
```diff
@@ -1039,6 +1044,52 @@ export const fetchFHIR = <T = any>(
     })
 }
 
+export const fetchFhirWithHearth = <T = any>(
```
[the remark this answers is not in the document]
Comment (kayumuzzaman) [reply packages/gateway/src/features/fhir/utils.ts]: fixed

File: packages/gateway/src/features/restLocation/locationHandler.ts [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    response = await fetchFhirWithHearth(`/Location${searchParam}`)
+  }
+
+  if (!response) {
+    throw internal('No location has been found')
+  } else {
+    return {
+      location: response
+    }
+  }
```
[the remark this answers is not in the document]
Comment (kayumuzzaman) [reply packages/gateway/src/features/restLocation/locationHandler.ts]: removed

File: packages/gateway/src/features/restLocation/locationHandler.ts (hunk 4 above)
Comment (kayumuzzaman) [reply packages/gateway/src/features/restLocation/locationHandler.ts]: fixed

File: packages/gateway/src/features/fhir/utils.ts [hunk 8]
```diff
@@ -1039,6 +1044,52 @@ export const fetchFHIR = <T = any>(
...
+  })
+    .then((response) => {
+      return response
+    })
+    .catch((error) => {
+      return Promise.reject(
+        new Error(`FHIR ${method} failed: ${error.message}`)
+      )
+    })
+}
```
[the remark this answers is not in the document]
Comment (kayumuzzaman) [reply packages/gateway/src/features/fhir/utils.ts:1091]: arise issue with response, so used `sendToFhir` instead of `fetchFHIR`

**Review by Zangetsu101 — APPROVED** [review]

**Comment by Zangetsu101** [comment]
@euanmillar The changes look good to me. Just give it a second look and we can merge this in

**Comment by euanmillar** [comment]
Great work @kayumuzzaman !  Lets merge this.

**Review by euanmillar — APPROVED** [review]

**Pull request merged**